tests/test_ensemble_tools.py

```diff
@@ -0,0 +1,239 @@
+            raise ValueError(msg)
+        key = match.group("key")
+        value = float(match.group("value"))
+        if key == parameter_name:
+            if expected_value == pytest.approx(value):
+                n_equals += 1
+            else:
+                n_different += 1
+
+    return (n_equals, n_different)
+
+
+@pytest.mark.parametrize("n_dirs", [1, 3, 5, 10])
+@pytest.mark.parametrize("destructive", [True, False])
+@pytest.mark.parametrize(
+    "parameter_and_range",
+    [
+        ["particle_initial_velocity", -4.0, 5.0],
+        ["particle_charge_density", 101.0, 108.0],
+        ["particle_number_density", 101.0, 108.0],
+    ],
+)
+def test_create_dir_tree(
+    tmp_path: Path,
+    *,
+    n_dirs: int,
+    destructive: bool,
+    parameter_and_range: list,
+) -> None:
+    """Test the create_dir_tree method of ensemble_tools."""
+    test_sweep_dir = tmp_path / "test_sweep"
+    copy_dir = Path("config_files") / "two_stream"
+    edit_file = "conditions.xml"
+    parameter_to_scan = parameter_and_range[0]
+    scan_range = (parameter_and_range[1], parameter_and_range[2])
+    outdir_prefix = "asd"
+
+    create_dir_tree(
+        sweep_path=test_sweep_dir,
+        n_dirs=n_dirs,
+        destructive=destructive,
+        copy_dir=copy_dir,
+        edit_file=edit_file,
+        parameter_to_scan=parameter_to_scan,
+        scan_range=scan_range,
+        outdir_prefix=outdir_prefix,
+    )
+
+    assert (test_sweep_dir / "SWEEP").is_dir()
+    assert len(list((test_sweep_dir / "SWEEP").iterdir())) == n_dirs
+    for i in range(n_dirs):
+        cond_file = test_sweep_dir / "SWEEP" / f"{outdir_prefix}{i}" / "conditions.xml"
+        assert cond_file.is_file()
+        para_value = (
+            scan_range[0]
+            if n_dirs == 1
+            else scan_range[0] + (i / (n_dirs - 1)) * (scan_range[1] - scan_range[0])
+        )
+        # Check the parameter has been edited correctly and appears only once
+        assert _check_parameter_in_conditions(
+            cond_file, parameter_to_scan, para_value
+        ) == (1, 0)
+        # Use an incorrect value of the parameter to get (0,1)
+        assert _check_parameter_in_conditions(
+            cond_file, parameter_to_scan, para_value + 10.0
+        ) == (0, 1)
+        # Make up a parameter and check we get (0,0)
+        assert _check_parameter_in_conditions(
+            cond_file, "fake_parameter", para_value
+        ) == (0, 0)
+
+
+def _call_create_dir_tree_from_dict(input_dict: dict) -> None:
+    create_dir_tree(
+        sweep_path=input_dict["sweep_path"],
+        n_dirs=input_dict["n_dirs"],
+        destructive=input_dict["destructive"],
+        copy_dir=input_dict["copy_dir"],
+        edit_file=input_dict["edit_file"],
+        parameter_to_scan=input_dict["parameter_to_scan"],
+        scan_range=input_dict["scan_range"],
+        outdir_prefix=input_dict["outdir_prefix"],
+    )
```

Comment:
So this is exactly what I was trying to originally - each of the `_call_create_dir_tree_from_dict(argument_dict)` calls were `create_dict_tree(**argument_dict)` - but `mypy` complained a lot about the type hinting for `create_dir_tree` not properly matching the dict arguments passed. I got the following errors for each instance:

```
tests/test_ensemble_tools.py:158: error: Argument 1 to "create_dir_tree" has incompatible type "**dict[str, object]"; expected "Path"  [arg-type]
tests/test_ensemble_tools.py:158: error: Argument 1 to "create_dir_tree" has incompatible type "**dict[str, object]"; expected "int"  [arg-type]
tests/test_ensemble_tools.py:158: error: Argument 1 to "create_dir_tree" has incompatible type "**dict[str, object]"; expected "bool"  [arg-type]
tests/test_ensemble_tools.py:158: error: Argument 1 to "create_dir_tree" has incompatible type "**dict[str, object]"; expected "str"  [arg-type]
tests/test_ensemble_tools.py:158: error: Argument 1 to "create_dir_tree" has incompatible type "**dict[str, object]"; expected "tuple[float, float]"  [arg-type]
Found 5 errors in 1 file (checked 1 source file)

``` 

I tried explicitly casting the objects for each keyword in `argument_dict`, but it didn't help. To avoid writing out the entire call each time I then refactored into the `_call_create_dir_tree_from_dict` method. If you know a way to avoid these errors in `mypy` I'd be happy to put it back!

Reply:
Ah I'd forgotten `mypy` is strict about this. One option would be to create a `TypedDict` instance, something like
```Python
class _CreateDirTreeArgsDict(TypedDict):
    sweep_path: Path
    n_dirs: int
    destructive: bool
    copy_dir: Path
    edit_file: str
    parameter_to_scan: str
    scan_range: tuple[float, float]
    outdir_prefix: str
```
and then swap `argument_dict` definition to
```Python
    argument_dict = _CreateDirTreeArgsDict(
        sweep_path=tmp_path / "test_sweep",
        n_dirs=5,
        destructive=True,
        copy_dir=Path("config_files") / "two_stream",
        edit_file="conditions.xml",
        parameter_to_scan="particle_initial_velocity",
        scan_range=(0.2, 5.0),
        outdir_prefix="asd",
    )
```
which should stop `mypy` complaining when unpacking. I think at the moment this would still leave an error when the `parameter_to_scan` argument is set to `None` as the type annotation for that argument says it should be a string rather than `None` type or a `str` (either wrap in `Optional` or `Union[..., None]`).

Reply:
Thanks, that all sounds sensible